chat-plugins/uno.js

```diff
@@ -0,0 +1,694 @@
+}
+
+function getUserName(id) {
+	return Users(id) ? Users(id).name : id;
+}
+
+class Game {
+	constructor(room, playerCap) {
+		if (!room.unoGameId) {
+			room.unoGameId = 0;
+		}
+		room.unoGameId++;
+		this.room = room;
+		this.id = 0;
+		this.top = null;
+		this.discard = [];
+		this.deck = null;
+		//for player order
+		this.list = [];
+		this.data = {};
+		this.joinedIps = {};
+		this.player = null;
+		this.gameid = "uno";
+		this.started = false;
+		this.timer = null;
+		this.lastDraw = null;
+		this.colourChange = null;
+		this.uhtmlChange = null;
+		this.playerCap = playerCap ? playerCap : null;
+	}
+
+	init() {
+		this.room.add("|uhtml|new" + this.room.unoGameId + "|<div class=\"broadcast-green\"><center><img src=\"http://www.theboardgamefamily.com/wp-content/uploads/2010/12/uno-mobile-game1.jpg\" height=150 width=160><br><font color=white><b>A new game of UNO is starting</b></font><br><button style=\"height: 25px ; width: 50px ;\" name=\"send\" value=\"/uno join\">Join</button></center></div>");
+	}
+
+	join(user) {
+		if (this.started) return false;
+		let ip = user.latestIp;
+		let userid = user.userid;
+
+		if (this.playerCap && this.list.length >= this.playerCap) {
+			return user.sendTo(this.room, "The game is already full.");
+		}
+
+		if (userid in this.data || ip in this.joinedIps) {
+			return user.sendTo(this.room, "You have already have an alt joined.");
+		}
+
+		this.list.push(userid);
+		//init empty hand
+		this.data[userid] = [];
+		this.joinedIps[ip] = 1;
+		//make join messages clear away after game starts
+		this.room.add("|uhtml|init" + this.room.unoGameId + "|" + user.name + " has joined the game.");
+	}
+
+	start(user) {
+		if (this.started) return false;
+		if (this.list.length < 2) return user.sendTo(this.room, "There aren't enough players!");
+		this.started = true;
+		this.room.add("The game has started!");
+		//hide the start message
+		this.room.add("|uhtmlchange|new" + this.room.unoGameId + "|<div class=\"infobox\">The game has started.</div>");
+		//get first player
+		this.player = this.list[~~(Math.random() * this.list.length)];
+		this.room.add("The first player is: " + getUserName(this.player) + ".");
+
+		//create the deck
+		this.deck = shuffleDeck(initDeck(this.list.length));
+
+		//deal the cards
+		this.list.forEach(function (u) {
+			this.giveCard(u, 7);
+		}.bind(this));
+
+		//get the top card and put it into discard pile
+		while (!this.top || this.top.charAt(0) === "Z") {
+			this.top = this.deck.shift();
```

Comment:
`shift()` removes an element from the *bottom* of the deque. Use `pop()` when you want the top